onnxruntime/test/optimizer/graph_transform_test.cc

```diff
@@ -7764,5 +7765,99 @@ TEST_F(GraphTransformationTests, ShapeInputMerge) {
                                         1, pre_graph_checker, post_graph_checker));
 }
 
+#if !defined(DISABLE_CONTRIB_OPS)
+
+TEST_F(GraphTransformationTests, MatMulNBitsBiasFusion) {
+  struct TestOptions {
+    bool bias_is_first_add_input{false};
+    bool add_produces_graph_output{false};
+  };
+
+  auto run_test = [&logger = *logger_](const TestOptions& opts) {
+    SCOPED_TRACE(MakeString("bias_is_first_add_input:", opts.bias_is_first_add_input,
+                            ", add_produces_graph_output:", opts.add_produces_graph_output));
+
+    auto build_test_case = [&](ModelTestBuilder& builder) {
+      constexpr size_t qbits = 4;
+      constexpr size_t block_size = 32;
+
+      constexpr int64_t M = 2, K = 4, N = 8;
+
+      int q_rows, q_cols;
+      MlasBlockwiseQuantizedShape<float, qbits>(block_size, /* columnwise */ true,
+                                                K, N,
+                                                q_rows, q_cols);
+
+      size_t q_data_size_in_bytes, q_scale_size, q_zp_size_in_bytes;
+      MlasBlockwiseQuantizedBufferSizes(qbits, block_size, /* columnwise */ true,
+                                        K, N,
+                                        q_data_size_in_bytes, q_scale_size, &q_zp_size_in_bytes);
+
+      auto* A = builder.MakeInput<float>(std::vector{M, K}, "A");
+
+      auto* B_data = builder.MakeInitializer<uint8_t>({int64_t{q_rows}, int64_t{q_cols}},
+                                                      uint8_t{0}, uint8_t{255});
+      auto* B_scales = builder.MakeInitializer<float>({static_cast<int64_t>(q_scale_size)},
+                                                      1.0f, 2.0f);
+      auto* B_zero_points = builder.MakeInitializer<uint8_t>({static_cast<int64_t>(q_zp_size_in_bytes)},
+                                                             uint8_t{0}, uint8_t{255});
+
+      auto* matmul_output = builder.MakeIntermediate();
+
+      auto& matmul = builder.AddNode("MatMulNBits",
+                                     {A, B_data, B_scales, B_zero_points},
+                                     {matmul_output},
+                                     kMSDomain);
+      matmul.AddAttribute("N", N);
+      matmul.AddAttribute("K", K);
+      matmul.AddAttribute("block_size", static_cast<int64_t>(block_size));
+      matmul.AddAttribute("bits", static_cast<int64_t>(qbits));
+
+      auto* Bias = builder.MakeInput<float>(std::vector{N}, "Bias");
+
+      auto* graph_output = builder.MakeOutput();
+
+      auto* add_output = opts.add_produces_graph_output ? graph_output : builder.MakeIntermediate();
+
+      builder.AddNode("Add",
+                      {opts.bias_is_first_add_input ? Bias : matmul_output,
+                       opts.bias_is_first_add_input ? matmul_output : Bias},
+                      {add_output});
+
+      if (!opts.add_produces_graph_output) {
+        builder.AddNode("Identity",
```

Comment:
>Identity [](http://example.com/codeflow?start=25&length=8)

does it work if the output of Add is also the output of model/graph

Reply:
it should. that is what is tested when `opts.add_produces_graph_output` is true